Show tags, import by drag-and-drop and avoid refreshed in workflow page

mvdbeek opened the pull request:
This adds support for tagging workflows directly on the overview page,
allows downloading, renaming and copying workflows without refreshing the page
and allows dragging workflow files into the center frame to import them.

This workflow also includes #4500, once that is merged it should be clearer.


![](https://github.com/mvdbeek/galaxy/raw/990906333d8fde3bb738a8f47e4756796fad1674/new_wf_menu_features.gif)

mvdbeek committed:
WIP - show/add/remove workflow tags

mvdbeek committed:
Allow editing tags in workflow overview page

guerler commented on client/galaxy/scripts/mvc/tag.js:
```diff
@@ -133,7 +135,70 @@ var TagsEditor = Backbone.View
 
 // =============================================================================
+
+var WorkflowTagsEditor = TagsEditor.extend({
```
Any chance we can re-use or extend the tags editor without introducing a workflow specific tag editor? I am asking because tag editors will be needed for other grids in the future too.

mvdbeek replied:
Yeah, I guess we could change the behavior of the tag editor (it's used for datasets and collections as well) to be the same as the WorkflowTagsEditor (i.e click on the tag to edit it, and close the editor with Escape instead of a separate button for opening/closing the editor).

guerler replied:
Yeah or you can add flags to govern the closing behavior or any other preferences, we have something similar here: 
https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/mvc/form/form-view.js#L12
This would avoid inheritance and would let us use combinations of options. It would be nice to have one customizable tag editor we can extend in this fashion and use throughout the ui.

mvdbeek replied:
@guerler I merged the WorkflowTagsEditor with the TagsEditor in 3ca147f, is that what you had in mind ?

mvdbeek committed:
Allow exporting and importing workflow tags

mvdbeek committed:
Allow updating workflows from partial information

The workflow overview page gets workflow information from the
/api/workflows, which returns a reduced amount of information, most
notably the steps are not being returned. This commit skips updating the
actual workflow steps if the steps are not included in the workflow
payload. We can now update name, tags and whether a workflow should be
shown in the tool panel. We also return owner and number of steps in the
response to importing workflows.

mvdbeek committed:
Remove unncessary boilerplate from workflow model

mvdbeek committed:
Allow renaming, downloading, dropping workflow files to import

mvdbeek commented:
We can now download and rename workflows without leaving the overview page or refreshing the whole page, and we can drop (one or many) workflow files to (batch) import them, but there is still some work left do.

mvdbeek committed:
Only access tags if workflow exists

mvdbeek committed:
Allow downloading .ga workflows via the UI

mvdbeek committed:
Remove downloading workflows through controller

mvdbeek committed:
Properly raise Exception if tool not found

mvdbeek committed:
Copy workflows without reloading page

mvdbeek committed:
Merge WorkflowTagsEditor with TagsEditor

mvdbeek committed:
Make tag_editor local variable

mvdbeek committed:
Fix logic for determining whether username should be mentioned

mvdbeek committed:
Fix adjusting dropdown menu

mvdbeek committed:
Use toastr for informing user about success/failure

mvdbeek committed:
Highlight workflow dropzone

mvdbeek committed:
Remove unused $el_wf_link variable

mvdbeek committed:
camelCase functions and variables

mvdbeek committed:
Display errors to user

jmchilton commented:
I missed this at first - I thought it was just some small tag change - this is so incredible! This eliminates so many clicks. I've merged your other PR, can you rebase to resolve conflicts?

martenson commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -1,94 +1,239 @@
+        initialize: function(){
+            _.bindAll(this, 'render', '_rowTemplate', 'renderTagEditor', '_templateActions', 'removeWorkflow', 'copyWorkflow'); // every function that uses 'this' as the current object should be in here
+            mod_toastr.options.timeOut = 1500;
+            mod_toastr.options.positionClass = "toast-top-left";
```
Could we please stick to top-right? I think all other implementations show notifications there.

mvdbeek replied:
Sure, np.

martenson commented on client/galaxy/scripts/mvc/workflow/workflow.js:
```diff
@@ -1,94 +1,239 @@
+        },
+
+        /** Template for user actions for workflows */
+        _templateActions: function( ) {
```
I think using the underscore templating would be helpful here but not necessary.

mvdbeek replied:
Can you point me to an example in the codebase ?

martenson replied:
https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/mvc/history/history-view-edit.js#L562

https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/mvc/library/library-librarytoolbar-view.js#L182

(I did not even know Carl wrote a [helper](https://github.com/galaxyproject/galaxy/blob/dev/client/galaxy/scripts/mvc/base-mvc.js#L555) for that template wrapping)

martenson commented on client/galaxy/scripts/mvc/workflow/workflow-model.js:
```diff
@@ -0,0 +1,42 @@
+
+var logNamespace = 'workflow';
+//==============================================================================
+/** @class model for tool citations.
```
copypasta?

mvdbeek replied:
indeed

martenson commented on client/galaxy/scripts/mvc/workflow/workflow-model.js:
```diff
@@ -0,0 +1,42 @@
+    url: '/api/workflows',
+
+    initialize: function () {
+        this.fetch();
```
I am not sure this belongs to the model, would it be better served in view with the rest of the controller's logic?

mvdbeek replied:
Makes sense, I'll have a look

martenson reviewed (commented):
Looks great. Few comments inline. It would be nice to also show only one notification when importing multiple workflows but that is minor.

mvdbeek committed:
Show toast at the top-right (which is the default setting)

jmchilton committed:
More workflow index page Selenium tests.

- Fix rename test for recent change to using alert.
- Test basic adding a tag.
- Outline of a test for downloading workflows.
- Add test for publishing display on index.
- Add test for using the search box on the workflow index page.

mvdbeek committed:
Describe workflow model and collection (thx @martenson)

mvdbeek committed:
Move fetching out of WorkflowCollection

thx for the hint @martenson.

mvdbeek committed:
Improve error reporting to user

mvdbeek committed:
Add static JS

jmchilton commented:
@mvdbeek I pushed some tests into your branch - hope that is cool - feel free to strip them out if not.

martenson commented:
@mvdbeek once you OK the tests that @jmchilton added I will merge. This is splendid.

mvdbeek commented:
Yeah, of course, tests are always awesome! Thanks @jmchilton and
@martenson. I'll keep the template changes for another PR.

On Sep 1, 2017 6:03 PM, "Martin Cech" <notifications@github.com> wrote:

> @mvdbeek <https://github.com/mvdbeek> once you OK the tests that
> @jmchilton <https://github.com/jmchilton> added I will merge. This is
> splendid.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/galaxyproject/galaxy/pull/4476#issuecomment-326619222>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AGfVpZ6-Wtfp4EiQTiMdD-_yCPHIlBamks5seCqzgaJpZM4O-zuC>
> .
>

martenson commented:
@mvdbeek your templating approach is not wrong, it is more of personal preference at this use case I guess. For a more complex template the underscore templating is a better choice I believe.
